google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatcherImpl.java

```diff
@@ -15,102 +15,87 @@
 
-    public Factory(TopicPath topicPath, AdminClient adminClient, Duration period) {
+    public Factory(TopicPath topicPath, AdminClient adminClient, AlarmFactory alarmFactory) {
       this.topicPath = topicPath;
       this.adminClient = adminClient;
-      this.period = period;
+      this.alarmFactory = alarmFactory;
     }
 
     @Override
     public PartitionCountWatcher newWatcher(Consumer<Long> receiver) {
-      return new PartitionCountWatcherImpl(topicPath, adminClient, receiver, period);
+      return new PartitionCountWatcherImpl(topicPath, adminClient, alarmFactory, receiver);
     }
   }
 
   private PartitionCountWatcherImpl(
-      TopicPath topicPath, AdminClient adminClient, Consumer<Long> receiver, Duration period) {
-    this.period = period;
+      TopicPath topicPath,
+      AdminClient adminClient,
+      AlarmFactory alarmFactory,
+      Consumer<Long> receiver) {
     this.topicPath = topicPath;
     this.adminClient = adminClient;
+    this.alarmFactory = alarmFactory;
     this.partitionCountReceiver = receiver;
   }
 
-  private void pollTopicConfig() {
-    Long partitionCount;
+  private void onAlarm() {
     try {
-      partitionCount = adminClient.getTopicPartitionCount(topicPath).get();
-    } catch (InterruptedException | ExecutionException e) {
-      // If we encounter an exception on our first topic config poll, then fail. We need to fetch
-      // the topic
-      // config at least once to start up properly.
-      if (!currentPartitionCount.isPresent()) {
-        notifyFailed(ExtractStatus.toCanonical(e));
-        stop();
-      }
-      log.atWarning().withCause(e).log("Failed to refresh partition count");
-      return;
-    }
-    if (currentPartitionCount.isPresent() && currentPartitionCount.get().equals(partitionCount)) {
-      return;
+      pollTopicConfig();
+    } catch (Throwable t) {
+      // Failures to update the partition count are not actual failures since we would have already
+      // fetched the first config.
+      log.atWarning().withCause(t).log("Failed to refresh partition count");
     }
+  }
+
+  private void pollTopicConfig() {
     try {
+      long partitionCount = adminClient.getTopicPartitionCount(topicPath).get();
+      if (currentPartitionCount == partitionCount) return;
+      currentPartitionCount = partitionCount;
       partitionCountReceiver.accept(partitionCount);
     } catch (Throwable t) {
-      notifyFailed(ExtractStatus.toCanonical(t));
-      stop();
+      throw toCanonical(t).underlying;
     }
-    // Notify started after we successfully receive the config once.
-    if (!currentPartitionCount.isPresent()) {
-      notifyStarted();
-    }
-    currentPartitionCount = Optional.of(partitionCount);
-  }
-
-  private void stop() {
-    partitionCountPoll.cancel(true);
-    adminClient.close();
   }
 
   @Override
   protected void doStart() {
-    partitionCountPoll =
-        SystemExecutors.getAlarmExecutor()
-            .scheduleAtFixedRate(
```

Comment:
This is one footgun that means there can be a data race if one of the poll requests locks up.